### PR TITLE
Feat/alerts scoped to env

### PR DESCRIPTION
**Pull request by marcambier**
**Issue**

https://github.com/gravitee-io/issues/issues/6079

**Description**

feat: scope platform alerts to environment

- Add a environment_id on each alert_trigger
- Platform alerts are scoped to environment, with referenceType=ENVIRONMENT & referenceId=environment.id

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-kepzbijuwt.chromatic.com)

🚀 CI was able to deploy the build of this PR, so you can now try it directly [here](https://apimnightlywebui24386.blob.core.windows.net/feat-alertsscopedtoenv-master/index.html)
_Notes_: The deployed app is linked to the management API of APIM master. (Same login and password as APIM master)

### CODE REVIEW

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/AlertServiceImpl.java [hunk 1]
```diff
@@ -357,10 +353,10 @@ public AlertTriggerEntity apply(AlertTrigger alertTrigger) {
     public void delete(final String alertId, final String referenceId) {
         try {
             final Optional<AlertTrigger> optionalAlert = alertTriggerRepository.findById(alertId);
-            if (!optionalAlert.isPresent() || !optionalAlert.get().getReferenceId().equals(referenceId)) {
+            if (optionalAlert.isEmpty() || !optionalAlert.get().getReferenceId().equals(referenceId)) {
                 throw new AlertNotFoundException(alertId);
             }
-            final AlertTriggerEntity alert = convert(optionalAlert.get());
+            final AlertTriggerEntity alert = alertTriggerConverter.toAlertTriggerEntity(optionalAlert.get());
```
Comment (a-cordier) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/AlertServiceImpl.java]: maybe even

```java
final AlertTriggerEntity alert = alertTriggerRepository.findById(alertId)
    .filter(a -> a.getReferenceId().equals(referenceId))
    .map(alertTriggerConverter::toAlertTriggerEntity)
    .orElseThrow(() -> new AlertNotFoundException(alertId));
```

File: .circleci/config.yml [hunk 2]
```diff
@@ -290,6 +290,9 @@ jobs:
             - run:
                   name: Run tests
                   command: |
+                      cd gravitee-apim-repository/gravitee-apim-repository-api
+                      mvn -s ../../.gravitee.settings.xml clean install -DskipTests
+                      cd ../..
```
Comment (gcusnieux) [inline .circleci/config.yml]:
```
- run:
      name: Build repository models
      command: |
            mvn clean install -s .gravitee.settings.xml 
            -f gravitee-apim-repository/gravitee-apim-repository-api  
            -Dskip.validation=true -DskipTests
- run:
      name: Run tests
      command: |
            mvn -pl 'gravitee-apim-definition, gravitee-apim-rest-api, gravitee-apim-gateway' -amd 
            --fail-fast -s .gravitee.settings.xml test --no-transfer-progress 
            -Dskip.validation=true -T 2C
```

File: gravitee-apim-console-webui/src/management/management.route.ts [hunk 3]
```diff
@@ -127,8 +127,9 @@ function managementRouterConfig($stateProvider) {
       controllerAs: '$ctrl',
       resolve: {
         configuredAlerts: (AlertService: AlertService) =>
-          AlertService.listAlerts(undefined, Scope.PLATFORM, false).then((response) => response.data),
-        alertingStatus: (AlertService: AlertService) => AlertService.getStatus(undefined, Scope.PLATFORM).then((response) => response.data),
+          AlertService.listAlerts(undefined, Scope.ENVIRONMENT, false).then((response) => response.data),
+        alertingStatus: (AlertService: AlertService) =>
+          AlertService.getStatus(undefined, Scope.ENVIRONMENT).then((response) => response.data),
       },
```
Comment (gcusnieux) [inline gravitee-apim-console-webui/src/management/management.route.ts:132]: Not related to your changes, but AlertService signatures could be refactored to pass the undefined as an optional parameter

File: gravitee-apim-rest-api/gravitee-apim-rest-api-portal/gravitee-apim-rest-api-portal-rest/src/main/java/io/gravitee/rest/api/portal/rest/resource/ApplicationAlertResource.java [hunk 4]
```diff
@@ -83,14 +83,18 @@ public Response updateAlert(
...
+            updateAlertTriggerEntity
+        );
 
         Alert alert = alertMapper.convert(updated);
         return Response.ok(alert).build();
     }
 
     private void checkPlugins() {
-        AlertStatusEntity alertStatus = applicationAlertService.getStatus();
+        AlertStatusEntity alertStatus = applicationAlertService.getStatus(GraviteeContext.getExecutionContext());
```
Comment (gcusnieux) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-portal/gravitee-apim-rest-api-portal-rest/src/main/java/io/gravitee/rest/api/portal/rest/resource/ApplicationAlertResource.java]: Could you pass the execution context as a parameter to instantiate it only once ?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-portal/gravitee-apim-rest-api-portal-rest/src/main/java/io/gravitee/rest/api/portal/rest/resource/ApplicationAlertsResource.java [hunk 5]
```diff
@@ -119,7 +119,7 @@ public ApplicationAlertResource getApplicationAlertResource() {
     }
 
     private void checkPlugins() {
-        AlertStatusEntity alertStatus = applicationAlertService.getStatus();
+        AlertStatusEntity alertStatus = applicationAlertService.getStatus(GraviteeContext.getExecutionContext());
```
Comment (gcusnieux) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-portal/gravitee-apim-rest-api-portal-rest/src/main/java/io/gravitee/rest/api/portal/rest/resource/ApplicationAlertsResource.java]: here also

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/converter/AlertTriggerConverter.java [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+                    .stream()
+                    .map(alertEventRuleEntity -> new AlertEventRule(AlertEventType.valueOf(alertEventRuleEntity.getEvent().toUpperCase())))
+                    .collect(toList())
+            );
+        }
+    }
+
+    public AlertTrigger toAlertTrigger(ExecutionContext executionContext, final UpdateAlertTriggerEntity alertEntity) {
+        final AlertTrigger alert = new AlertTrigger();
+        alert.setId(UUID.toString(UUID.random()));
```
Comment (gcusnieux) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/converter/AlertTriggerConverter.java]: This should not be `alert.setId(alertEntity.getId())` ?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/AlertServiceImpl.java [hunk 7]
```diff
@@ -152,45 +152,48 @@ public class AlertServiceImpl extends TransactionalService implements AlertServi
     private ApiRepository apiRepository;
 
     @Autowired
-    private ApiMetadataService apiMetadataService;
+    private EnvironmentService environmentService;
 
     @Override
-    public AlertStatusEntity getStatus() {
+    public AlertStatusEntity getStatus(ExecutionContext executionContext) {
```
Comment (gcusnieux) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/AlertServiceImpl.java]: We could use final ExecutionContent on each signatures

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/AlertServiceImpl.java [hunk 8]
```diff
@@ -410,102 +403,48 @@ public AlertEventEntity apply(AlertEvent alertEventRepo) {
...
+                                    AlertTriggerEntity triggerEntity = alertTriggerConverter.toAlertTriggerEntity(trigger);
+                                    triggerEntity.setId(UUID.toString(UUID.random()));
+                                    triggerEntity.setReferenceType(API);
+                                    triggerEntity.setReferenceId(apiId);
+                                    triggerEntity.setTemplate(false);
+                                    triggerEntity.setEnabled(true);
+                                    triggerEntity.setEventRules(null);
+                                    triggerEntity.setParentId(alertId);
+                                    triggerEntity.setCreatedAt(new Date());
+                                    triggerEntity.setUpdatedAt(trigger.getCreatedAt());
```
Comment (gcusnieux) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/AlertServiceImpl.java:441]: Not update by you, but it's strange no ?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-portal/gravitee-apim-rest-api-portal-rest/src/main/java/io/gravitee/rest/api/portal/rest/resource/ApplicationAlertResource.java (hunk 4 above)
Comment (marcambier) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-portal/gravitee-apim-rest-api-portal-rest/src/main/java/io/gravitee/rest/api/portal/rest/resource/ApplicationAlertResource.java]: :+1:

File: gravitee-apim-rest-api/gravitee-apim-rest-api-portal/gravitee-apim-rest-api-portal-rest/src/main/java/io/gravitee/rest/api/portal/rest/resource/ApplicationAlertsResource.java (hunk 5 above)
Comment (marcambier) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-portal/gravitee-apim-rest-api-portal-rest/src/main/java/io/gravitee/rest/api/portal/rest/resource/ApplicationAlertsResource.java]: :+1:

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/converter/AlertTriggerConverter.java (hunk 6 above)
Comment (marcambier) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/converter/AlertTriggerConverter.java]: Yes.
It was already like that in the mapping before refactoring.
But it's nonsense, the caller was setting back the id using the one from UpdateAlertTriggerEntity.
I changed it. Nice catch.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/AlertServiceImpl.java (hunk 1 above)
Comment (marcambier) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/AlertServiceImpl.java]: nice, thanks :+1:

File: gravitee-apim-console-webui/src/management/management.route.ts (hunk 3 above)
Comment (marcambier) [reply gravitee-apim-console-webui/src/management/management.route.ts:132]: Yes, I refactored that part in a dedicated commit :+1:

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/AlertServiceImpl.java (hunk 8 above)
Comment (marcambier) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/AlertServiceImpl.java:441]: Yes.
The updatedAt takes the creation date of the template (so, update date is before creation date).
I Can't explain why.

File: .circleci/config.yml (hunk 2 above)
Comment (marcambier) [reply .circleci/config.yml]: This is not related to this PR.
Will apply your remark on a dedicated PR.

**Review by gcusnieux — APPROVED** [review]

**Review by ThibaudAV — APPROVED** [review]
LGTM 👏 👏 

Test locally and sounds good :)

**Pull request merged**